### ISSUE TITLE
How to implement alternate access credentials.

### ISSUE THREAD

**Issue by macsolving** [Simbul/baker#913]
I'm trying to allow authorized access to content from a remote database of users in _addition_ to the apple IAP system.  I was wondering if someone could save me from waddling through a ton of code to figure out exactly where to implement this.  Currently I'm going at it from the following angle:  Edit    retrievePurchasesFor:productIDs   in PurchaseManager.m to check for subscription and issue access, and build purchasesResponse in that method as opposed to getting it from getPurchasesJSON, then proceed as usual.  Does anyone see any blatant problems with that solution, or have an alternate suggestion of a different direction to point me in.  Many Thanks!!!!

**Comment by folletto** [comment]
It should be doable without changing the code: when the app contacts the server, simply return the issues that are unlocked, no? :)

**Comment by macsolving** [comment]
This had occurred to me and may well be the easiest implementation -- we are using magrocket to handle the subscription/purchase info (although we're not tied down to it), so it would require some modification on that end.  Mostly I was afraid that magrocket would invalidate access info if it couldn't be verified with apple.

**Comment by nin9creative** [comment]
@macsolving You could always change some of the API code on the MagRocket side if you wanted to tweak things...  At some point in the future I will be adding the ability to provide a mechanism for what you are discussing (purchases / subscriptions) made outside of IAP.  However at this point you could probably add a database table or do some other database work on MagRocket, then just tweak the API code a bit and accomplish your goal.

**Comment by macsolving** [comment]
that's what i was thinking, maybe just add a column to the magrocket user database for username/id with the other member service (amember)?  nin9 do you see an issue with with what I mentioned about magrocket invalidating access if it can't be verified with apple?  (even if so, fixable I'm sure.....)

**Comment by nin9creative** [comment]
If you look at the API, there is a /purchases/ method that returns the list of available issues to Baker along with whether or not the user is subscribed.

All you'd need to do is change logic in there to also take whatever you want into consideration.

The problem with this in general is coming up with a way to match an end user's device with a purchase made outside Baker IAP.

Andrew

**Comment by macsolving** [comment]
awesome -- thank you

**Comment by macsolving** [comment]
ok Andrew - i'm noticing the magrocket user ID's appear to be hashed.  A couple of questions here, 1) is the userID hashed before it is passed to magrocket?  2) is the password passed with it?  Assuming magrocket verifies data with apple it would seem that answer to the former question is no and the latter is yes.  If so, are these parameters passed in the url?  As for matching a user's device, I think I can modify the API to pass the amember ID along with the apple ID, and marry that info, referencing the amember install in the purchases method as you suggest.  Do you see any obvious caveats in my login here?

**Comment by nin9creative** [comment]
That's what I was saying before.  The "USER_ID" that baker passes is nothing more than a generated unique identifier meant to identify the instance of Baker installed on a user's device.  Unless there is a mechanism within Baker to allow a user to enter some sort of additional data that you can use on the backend to match up with the USER_ID I don't know how this really can be done correctly.  I know that foletto made a comment earlier but I still don't know how you can identify the user on the backend without some sort of additional work on the Baker front end...

Let's say that I have a website where I sell subscriptions, or individual issues.  OR even that I want to give out promotional codes for subscriptions or issues.  You could code MagRocket to take those into consideration etc etc, but unless there is a front end UI where a user can enter the code, or provide some sort of credentials that will tie the USER_ID GUID (which is basically meaningless for identification purposes) to the issues or subscriptions that they have (that aren't run through IAP) I don't know how you could solve this issue.

**Comment by nin9creative** [comment]
I think this is basically the same issue as #886.

**Comment by macsolving** [comment]
I've got the interface coded in baker to retrieve an amember id/password -- so I think I should be able to modify the magrocket api to take that info at the same time as the USER_ID from baker, no?

**Comment by nin9creative** [comment]
Yeah, it's pretty easy to add another few parameters to the API definition...  Just look at the code and you'll pretty much understand how it works.

**Comment by macsolving** [comment]
Yeah, it's in /magrocket-api/index.php, right?

**Comment by nin9creative** [comment]
yes

**Comment by macsolving** [comment]
rockin -- thank you: hoping to get this implemented in the next couple days -- I'll give you an update

**Comment by macsolving** [comment]
Ok, so I'm set up returning content as purchased from magrocket (purchases method in /magrocket-api/index.php) when the secondary login service authorizes it, and the issues show up as purchased in baker, however, when I click on download, it says 'Download Failed'.  I'm assuming I need to specify download paths for the purchased content to baker, as well as indicating that it is purchased.  I'm stepping through baker to compare a sucessfull purchase via apple's IAP, and comparing to the failed download with the 'faked' purchase, but it's not immediately apparent where baker is failing to grab the download url.  Anybody have an idea for me?

**Comment by folletto** [comment]
Is the book in the `shelf.json` file?

**Comment by nin9creative** [comment]
Baker gets the available issues via the /issues/ method and then gets what the user has available with the /purchases/ method to
mash up what should be listed as purchased.

Andrew

**Comment by macsolving** [comment]
that's as I understood -- if I'm not mistaken the /issues/ api populates the baker shelf, and /purchases/ tells baker that a user has access to a issue:  I'm showing that the issues are purchased, but they aren't successfully downloading when they are shown as purchased using the modified /purchases/ method.  I'm assuming that the accurate download url isn't being passed to baker, but I'm spinning my wheels attempting to find where that is.  My hunch is that if I modify the magrocket database to show that a specific user has purchased a specific item, I will not get this problem - and that may ultimately be the solution, but I'd rather not interfere with magrocket's natural setup for keeping track of apple iap's because concievably a user could abandon the secondary subscription service for apple, or vice versa, and I'd rather not confuse the two....

**Comment by nin9creative** [comment]
The /issues/ api also contains a pointer to the download location.  That's how Baker knows where to get it from.  In MagRocket, it's defined as part of the Issues data.

**Comment by macsolving** [comment]
gotcha -- that's where i was headed, and yes, I see it getting called during the purchase in baker...

**Comment by macsolving** [comment]
if I were to modify the /issues/ or /purchases/ api to take both the login id and password for our secondary subscription service, would it be fairly easy to tweak the api to accept parameters as POST instead of GET?  From Baker I think I all I would have to do is change the NSURLRequest method to POST.  I'm not opposed to just implementing SSL certs and keeping the api calls using GET, but my understanding is the parameters would be recorded in the hosts connection logs

**Comment by nin9creative** [comment]
It's not hard to implement POST on the API, see the confirm purchase for example.

You could also implement SSL

Andrew

**Comment by macsolving** [comment]
SSL is cool -- the concern is that the login creds would be stored as url parameters in the hosts connection logs
-- but yeah, POST in combination with SSL should do the trick
